<a href="https://colab.research.google.com/github/Gaskillaaa/Gaskillaaa/blob/main/%D0%97%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5_%D0%BF%D0%BE_%D0%BE%D0%BF%D1%82%D0%B8%D0%BC%D0%B8%D0%B7%D0%B0%D1%86%D0%B8%D0%B8_%D0%B8_%D0%BF%D0%BE%D0%B8%D1%81%D0%BA%D1%83_%D0%B3%D0%B8%D0%BF%D0%B5%D1%80%D0%BF%D0%B0%D1%80%D0%B0%D0%BC%D0%B5%D1%82%D1%80%D0%BE%D0%B2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Загрузка и подготовка данных**

In [2]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

# Загрузка данных
df = pd.read_csv('/content/_train_sem09 (1).csv')
df

Activity        D1        D2    D3   D4        D5        D6        D7  \
0            1  0.000000  0.497009  0.10  0.0  0.132956  0.678031  0.273166   
1            1  0.366667  0.606291  0.05  0.0  0.111209  0.803455  0.106105   
2            1  0.033300  0.480124  0.00  0.0  0.209791  0.610350  0.356453   
3            1  0.000000  0.538825  0.00  0.5  0.196344  0.724230  0.235606   
4            0  0.100000  0.517794  0.00  0.0  0.494734  0.781422  0.154361   
...        ...       ...       ...   ...  ...       ...       ...       ...   
3746         1  0.033300  0.506409  0.10  0.0  0.209887  0.633426  0.297659   
3747         1  0.133333  0.651023  0.15  0.0  0.151154  0.766505  0.170876   
3748         0  0.200000  0.520564  0.00  0.0  0.179949  0.768785  0.177341   
3749         1  0.100000  0.765646  0.00  0.0  0.536954  0.634936  0.342713   
3750         0  0.133333  0.533952  0.00  0.0  0.347966  0.757971  0.230667   

            D8        D9  ...  D1767  D1768  D1769  D1770  D1771  D1772  \
0     0.585445  0.743663  ...      0      0      0      0      0      0   
1     0.411754  0.836582  ...      1      1      1      1      0      1   
2     0.517720  0.679051  ...      0      0      0      0      0      0   
3     0.288764  0.805110  ...      0      0      0      0      0      0   
4     0.303809  0.812646  ...      0      0      0      0      0      0   
...        ...       ...  ...    ...    ...    ...    ...    ...    ...   
3746  0.376124  0.727093  ...      0      0      0      0      0      0   
3747  0.404546  0.787935  ...      0      0      1      0      1      0   
3748  0.471179  0.872241  ...      0      0      0      0      0      0   
3749  0.447162  0.672689  ...      0      0      0      0      0      0   
3750  0.272652  0.854116  ...      0      0      0      0      0      0   

      D1773  D1774  D1775  D1776  
0         0      0      0      0  
1         0      0      1      0  
2         0      0      0      0  
3         0      0      0      0  
4         0      0      0      0  
...     ...    ...    ...    ...  
3746      0      0      0      0  
3747      1      0      0      0  
3748      0      0      0      0  
3749      0      0      0      0  
3750      0      0      0      0  

[3751 rows x 1777 columns]

In [3]:

# Разделение данных на X (признаки) и y (целевой столбец)
X = df.drop(columns=['Activity'])
y = df['Activity']

# Разделение на обучающую и тестовую выборки (80/20)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=42)

**Обучение базовых моделей**

Логистическая регрессия

In [4]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import f1_score

log_reg = LogisticRegression(random_state=42, max_iter=1000)
log_reg.fit(X_train, y_train)
y_pred = log_reg.predict(X_test)

print("F1-score (Logistic Regression):", f1_score(y_test, y_pred))

F1-score (Logistic Regression): 0.7773766546329723


Случайный лес

In [5]:
from sklearn.ensemble import RandomForestClassifier

rf = RandomForestClassifier(random_state=42, n_estimators=100)
rf.fit(X_train, y_train)
y_pred_rf = rf.predict(X_test)

print("F1-score (Random Forest):", f1_score(y_test, y_pred_rf))

F1-score (Random Forest): 0.8146341463414634


**Подбор гиперпараметров**

GridSearchCv

In [6]:
from sklearn.model_selection import GridSearchCV

param_grid = {
    'C': [0.01, 0.1, 1, 10, 100]
}

grid_search = GridSearchCV(LogisticRegression(max_iter=1000, random_state=42), param_grid, scoring='f1', cv=5)
grid_search.fit(X_train, y_train)

print("Лучшие параметры (GridSearchCV):", grid_search.best_params_)
print("Лучший F1-score:", grid_search.best_score_)

Лучшие параметры (GridSearchCV): {'C': 0.01}
Лучший F1-score: 0.7881060993690068


RandomizedSearchCV

In [13]:
from sklearn.model_selection import RandomizedSearchCV
from sklearn import linear_model
from sklearn import metrics
#np.linspace(start(от), stop(до), num=50(количество),dtype-тип данных)
param_distributions = {'penalty': ['l2', 'none'] ,
              'solver': ['lbfgs', 'sag'],
               'C': list(np.linspace(0.01, 1, 10, dtype=float))}

random_search = RandomizedSearchCV(
    estimator=linear_model.LogisticRegression(random_state=42, max_iter=1000),
    param_distributions=param_distributions,
    cv=5,
    n_iter = 10,
    n_jobs = -1
)


random_search.fit(X_train, y_train)
print("accuracy на тестовом наборе: {:.2f}".format(random_search.score(X_test, y_test)))
y_test_pred = random_search.predict(X_test)
print('f1_score на тестовом наборе: {:.2f}'.format(metrics.f1_score(y_test, y_test_pred)))
print("Наилучшие значения гиперпараметров: {}".format(random_search.best_params_))


/usr/local/lib/python3.11/dist-packages/sklearn/model_selection/_validation.py:528: FitFailedWarning: 
25 fits failed out of a total of 50.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
19 fits failed with the following error:
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/dist-packages/sklearn/model_selection/_validation.py", line 866, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/usr/local/lib/python3.11/dist-packages/sklearn/base.py", line 1382, in wrapper
    estimator._validate_params()
  File "/usr/local/lib/python3.11/dist-packages/sklearn/base.py", line 436, in _validate_params
    validate_parameter_constraints(
  File "/usr/local/lib/python3.11/dist-packages/sklearn/utils/_

accuracy на тестовом наборе: 0.76
f1_score на тестовом наборе: 0.78
Наилучшие значения гиперпараметров: {'solver': 'sag', 'penalty': 'l2', 'C': 0.01}


 Hyperopt (байесовская оптимизация)

In [14]:
from hyperopt import fmin, tpe, hp, Trials
from sklearn.model_selection import cross_val_score

def objective(params):
    model = RandomForestClassifier(
        n_estimators=int(params['n_estimators']),
        max_depth=int(params['max_depth']),
        min_samples_split=int(params['min_samples_split']),
        min_samples_leaf=int(params['min_samples_leaf']),
        random_state=42
    )
    score = cross_val_score(model, X_train, y_train, cv=5, scoring='f1').mean()
    return -score  # Hyperopt минимизирует функцию, поэтому берем отрицательное значение

space = {
    'n_estimators': hp.quniform('n_estimators', 50, 500, 50),
    'max_depth': hp.quniform('max_depth', 5, 50, 5),
    'min_samples_split': hp.quniform('min_samples_split', 2, 10, 1),
    'min_samples_leaf': hp.quniform('min_samples_leaf', 1, 5, 1),
}

trials = Trials()
best = fmin(fn=objective, space=space, algo=tpe.suggest, max_evals=30, trials=trials)
print("Лучшие параметры (Hyperopt):", best)

100%|██████████| 30/30 [09:02<00:00, 18.09s/trial, best loss: -0.819312814957047]
Лучшие параметры (Hyperopt): {'max_depth': 25.0, 'min_samples_leaf': 1.0, 'min_samples_split': 3.0, 'n_estimators': 100.0}


Optuna (оптимизация с байесовскими методами)

In [16]:
!pip install optuna

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 383.4/383.4 kB 11.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 233.6/233.6 kB 12.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 78.5/78.5 kB 5.1 MB/s eta 0:00:00


In [17]:
import optuna

def objective(trial):
    rf = RandomForestClassifier(
        n_estimators=trial.suggest_int("n_estimators", 50, 500, step=50),
        max_depth=trial.suggest_int("max_depth", 5, 50, step=5),
        min_samples_split=trial.suggest_int("min_samples_split", 2, 10),
        min_samples_leaf=trial.suggest_int("min_samples_leaf", 1, 5),
        random_state=42
    )
    score = cross_val_score(rf, X_train, y_train, cv=5, scoring="f1").mean()
    return score  # Optuna максимизирует значение

study = optuna.create_study(direction="maximize")
study.optimize(objective, n_trials=30)

print("Лучшие параметры (Optuna):", study.best_params)

[I 2025-02-04 15:47:19,088] A new study created in memory with name: no-name-576fe348-5b86-46a3-976b-fb81ee052ac0
[I 2025-02-04 15:47:46,764] Trial 0 finished with value: 0.7707159019185068 and parameters: {'n_estimators': 500, 'max_depth': 5, 'min_samples_split': 2, 'min_samples_leaf': 3}. Best is trial 0 with value: 0.7707159019185068.
[I 2025-02-04 15:48:30,030] Trial 1 finished with value: 0.8191450954019077 and parameters: {'n_estimators': 500, 'max_depth': 50, 'min_samples_split': 3, 'min_samples_leaf': 2}. Best is trial 1 with value: 0.8191450954019077.
[I 2025-02-04 15:48:57,427] Trial 2 finished with value: 0.8181309582997225 and parameters: {'n_estimators': 300, 'max_depth': 30, 'min_samples_split': 2, 'min_samples_leaf': 2}. Best is trial 1 with value: 0.8191450954019077.
[I 2025-02-04 15:49:31,574] Trial 3 finished with value: 0.8171869941115398 and parameters: {'n_estimators': 400, 'max_depth': 25, 'min_samples_split': 9, 'min_samples_leaf': 1}. Best is trial 1 with value:

Лучшие параметры (Optuna): {'n_estimators': 450, 'max_depth': 20, 'min_samples_split': 4, 'min_samples_leaf': 1}


# Вывод:
    1. Обучены базовые модели (логистическая регрессия и случайный лес)
    2. Подобраны гиперпараметры 4 разными методами
    3. Итоговый f1 отображен для каждой модели


Наилучшие показатели f1 получились используя Optima оптимизацию с параметрами n_estimators = 450, max depth = 20, min_samples_split = 4, min_samples_leaf = 1